### PR TITLE
Support multi duration 

### PR DESCRIPTION
**Pull request by rubio41**
Support multi duration such as "1 month 21 day"

### CODE REVIEW

File: .NET/Microsoft.Recognizers.Text.DateTime/Constants.cs [hunk 1]
```diff
@@ -22,5 +22,11 @@ public static class Constants
         public const string ResolveKey = "resolve";
         public const string ResolveToPastKey = "resolveToPast";
         public const string ResolveToFutureKey = "resolveToFuture";
+        public const string MutiDurationType = "mutiDurationType";
+
+        //Mutiple Duration Type
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Constants.cs]: // Multiple Duration Types

File: .NET/Microsoft.Recognizers.Text.DateTime/Constants.cs [hunk 2]
```diff
@@ -22,5 +22,11 @@ public static class Constants
         public const string ResolveKey = "resolve";
         public const string ResolveToPastKey = "resolveToPast";
         public const string ResolveToFutureKey = "resolveToFuture";
+        public const string MutiDurationType = "mutiDurationType";
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Constants.cs]: MultiDurationType and multiDurationType

File: .NET/Microsoft.Recognizers.Text.DateTime/Constants.cs [hunk 3]
```diff
@@ -22,5 +22,11 @@ public static class Constants
         public const string ResolveKey = "resolve";
         public const string ResolveToPastKey = "resolveToPast";
         public const string ResolveToFutureKey = "resolveToFuture";
+        public const string MutiDurationType = "mutiDurationType";
+
+        //Mutiple Duration Type
+        public const string MutiDuration_DateTime = "mutiDurationDateTime";
+        public const string MutiDuration_Date = "mutiDurationDate";
+        public const string MutiDuration_Time = "mutiDurationTime";
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Constants.cs]: Same

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateExtractor.cs [hunk 4]
```diff
@@ -181,11 +181,17 @@ private List<Token> DurationWithBeforeAndAfter(string text, DateObject reference
 
             foreach (var er in durationEr)
             {
-                var match = config.DateUnitRegex.Match(er.Text);
-                if (match.Success)
-                {
-                    ret = AgoLaterUtil.ExtractorDurationWithBeforeAndAfter(text,
-                        er, ret, config.UtilityConfiguration);
+                // if it is a muti-duration and its type is not equal to Date than skip it.
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDateExtractor.cs]: Same, and so on.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateParser.cs [hunk 5]
```diff
@@ -470,8 +470,7 @@ private DateTimeResolutionResult ParseDurationWithAgoAndLater(string text, DateO
                 config.DurationParser,
                 config.UnitMap,
                 config.UnitRegex,
-                config.UtilityConfiguration,
-                AgoLaterUtil.AgoLaterMode.Date
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateParser.cs]: Is removing this correct?

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDurationParser.cs [hunk 6]
```diff
@@ -30,9 +31,13 @@ public DateTimeParseResult Parse(ExtractResult er, DateObject refTime)
...
-
-                innerResult = ParseNumerWithUnit(er.Text, referenceTime);
+                var innerResult = new DateTimeResolutionResult();
+
+                innerResult = ParseMergedDuration(er.Text, referenceTime);
+                if (!innerResult.Success)
+                {
+                    innerResult = ParseNumerWithUnit(er.Text, referenceTime);
+                }
                 if (!innerResult.Success)
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDurationParser.cs]: Nit: add blank lines between ifs.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDurationParser.cs [hunk 7]
```diff
@@ -310,5 +315,78 @@ private bool TryGetResultFromRegex(Regex regex, string text, string numStr, out
...
+                        timexDict.Add(unitMatch.Groups["unit"].Value, pr.TimexStr);
+                        prs.Add(pr);
+                    }
+                }
+            }
+
+            if (prs.Count == ers.Count)
+            {
+                var unitList = new List<string>(timexDict.Keys);
+                unitList.Sort((x, y) => (this.config.UnitValueMap[x] < this.config.UnitValueMap[y]? 1 : -1));
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDurationParser.cs]: Can you add a code comment here explaining this sort?

File: Patterns/English/English-DateTime.yaml [hunk 8]
```diff
@@ -425,6 +425,8 @@ GeneralEndingRegex: !simpleRegex
   def: ^\s*((\.,)|\.|,|!|\?)?\s*$
 MiddlePauseRegex: !simpleRegex
   def: \s*(,)\s*
+DurationConnectorRegex: !simpleRegex
```
Comment (tellarin) [inline Patterns/English/English-DateTime.yaml]: Does this breaks generation of the .ts files for the Javascript port?

File: Specs/DateTime/English/DateParser.json [hunk 9]
```diff
@@ -2044,4 +2044,100 @@
     }
   ]
 }
+,
+{
+  "Input": "We had a meeting 1 month 21 days ago",
```
Comment (tellarin) [inline Specs/DateTime/English/DateParser.json]: It doesn't support "1 months and 21 days ago"?
Or "1 month, 21 days ago"?

Comment (tellarin) [reply Specs/DateTime/English/DateParser.json]: What about short forms like "1y 3d"?
And "one week three days"?

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateParser.cs (hunk 5 above)
Comment (rubio41) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateParser.cs]: yes. Determined the mode in the code before.

File: Specs/DateTime/English/DateParser.json (hunk 9 above)
Comment (rubio41) [reply Specs/DateTime/English/DateParser.json]: "1 months and 21 days ago" and "1 month, 21 days ago" are supported.
"one week three days" is supported.
"1y 3d" is not supported. We do not extract "1y" and year. The same for "3d".

Comment (tellarin) [reply Specs/DateTime/English/DateParser.json]: Can you add test cases for the 3 supported ones?

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDurationExtractor.cs [hunk 10]
```diff
@@ -101,6 +111,121 @@ private static List<Token> GetTokenFromRegex(Regex regex, string text)
...
+            {
+                return extractorResults;
+            }
+
+            var UnitMap = this.config.UnitMap;
+            var UnitValueMap = this.config.UnitValueMap;
+            var unitRegex = this.config.DurationUnitRegex;
+            List<ExtractResult> ret = new List<ExtractResult>();
+
+            var idx_i = 0;
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDurationExtractor.cs]: Let's give variables better names than idx_i and idx_j.

**Pull request merged**